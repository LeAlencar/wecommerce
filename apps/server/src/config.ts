import dotenvSafe from 'dotenv-safe';
import path from 'path';

const cwd = process.cwd();

const root = path.join.bind(cwd);

dotenvSafe.config({
  path: root('.env'),
  sample: root('.env.example')
});

const env = process.env;

export const database = env.MONGO_URI as string;
export const jwtSecret = env.JWT_SECRET as string;

export const auth = {
  user: env.SMTP_AUTH_USER as string,
  pass: env.SMTP_AUTH_PASS as string
};
export const service = env.EMAIL_SERVICE as string;
