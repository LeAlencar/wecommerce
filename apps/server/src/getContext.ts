import { IUser } from './modules/user/UserModel';
import { Context } from 'koa';
import { getDataloaders } from './modules/loader/loaderRegister';
import { GraphQLContext } from './types/types';
import { Request as HelixRequest } from 'graphql-helix';
import { load } from './modules/user/UserLoader';

type ContextVars = {
  user?: IUser | null;
  req?: HelixRequest;
  koaContext: Context;
  setCookie: (cookieName: string, token: string) => void;
};

export const getContext = async (ctx: ContextVars) => {
  const context = {
    ...ctx
  };
  const dataloaders = getDataloaders();

  if (context.user) {
    context.user = await load({ ...context, dataloaders }, context.user._id);
  }

  return {
    req: ctx.req,
    dataloaders,
    user: ctx.user,
    koaContext: ctx.koaContext,
    setCookie: ctx.setCookie
  } as GraphQLContext;
};
