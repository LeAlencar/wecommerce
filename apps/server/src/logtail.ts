import { Logtail } from '@logtail/node';

export default function getLogtailClient() {
  const logtailClient = new Logtail('ZqUp4cYzxXiS1eVPBdD5KSts');
  return logtailClient;
}
