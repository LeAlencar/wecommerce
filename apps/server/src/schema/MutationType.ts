import { GraphQLObjectType } from 'graphql';

import UserMutations from '../modules/user/mutations';
import ProductMutations from '../modules/product/mutations';
import ChargeMutations from '../modules/charge/mutations';

export default new GraphQLObjectType({
  name: 'Mutation',
  description: 'Root of ... mutations',
  fields: () => ({
    ...UserMutations,
    ...ProductMutations,
    ...ChargeMutations
  })
});
