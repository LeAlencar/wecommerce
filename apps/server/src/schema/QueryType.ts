import { GraphQLObjectType } from 'graphql';
import { UserType } from '../modules/user/UserType';
import { nodeField, nodesField } from '../modules/node/typeRegister';
import * as UserLoader from '../modules/user/UserLoader';

const QueryType = new GraphQLObjectType({
  name: 'Query',
  description: 'The root of all queries',
  fields: () => ({
    node: nodeField,
    nodes: nodesField,
    //...userConnectionField(),
    me: {
      type: UserType,
      resolve: async (_, args, context) => {
        return await UserLoader.load(context, context.user?._id);
      }
    }
  })
});
export default QueryType;
