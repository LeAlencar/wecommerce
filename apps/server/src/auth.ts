import jwt from 'jsonwebtoken';
import { jwtSecret } from './config';
import UserModel, { IUser } from './modules/user/UserModel';

export const getUser = async (token: string | null | undefined) => {
  if (!token) return { user: null };

  try {
    const decodedToken = jwt.verify(
      token?.replace('JWT', '').trim(),
      jwtSecret
    );

    const user = await UserModel.findOne({
      _id: (decodedToken as { id: string }).id
    });

    return {
      user
    };
  } catch (err) {
    console.log(err);
    return { user: null };
  }
};

export const generateJwtToken = (user: IUser) => {
  return `JWT ${jwt.sign({ id: user._id }, `jwtsupersecreto`)}`;
};
