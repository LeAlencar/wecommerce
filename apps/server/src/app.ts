import cors from '@koa/cors';
import Router from '@koa/router';
import {
  getGraphQLParameters,
  processRequest,
  renderGraphiQL,
  sendResult,
  shouldRenderGraphiQL
} from 'graphql-helix';
import Koa, { Context } from 'koa';
import bodyParser from 'koa-bodyparser';
import koaLogger from 'koa-logger';
import koaPlayground from 'graphql-playground-middleware-koa';
import { getContext } from './getContext';
import { schema } from './schema/schema';
import { getUser } from './auth';
import getLogtailClient from './logtail';

const app = new Koa();

const router = new Router();
const logtailClient = getLogtailClient();

export const statusMiddleware = async (ctx: Record<string, unknown>) => {
  try {
    ctx.body = 'Server Working';
    ctx.status = 200;
  } catch (err: unknown) {
    console.log({
      err
    });
    ctx.body = err.toString();
    ctx.status = 400;
  }
};

app.use(bodyParser());

app.use(koaLogger());
app.use(
  cors({
    maxAge: 86400,
    credentials: true,
    origin: '*'
  })
);

export const setCookie =
  (context: Context) => async (cookieName: string, token: string) => {
    context.cookies.set(cookieName, token, {
      domain:
        process.env.NODE_ENV === 'production' ? 'lealencar.dev' : undefined,
      httpOnly: true,
      secure: process.env.NODE_ENV == 'production' ? true : false,
      sameSite: 'lax',
      path: '/',
      maxAge: 365 * 24 * 60 * 60 * 100
    });
  };

router.get('/status', statusMiddleware);

router.all(
  '/playground',
  koaPlayground({
    endpoint: '/graphql'
  })
);

router.all('/graphql', async (ctx) => {
  const userToken = ctx.cookies.get('userToken');
  logtailClient.info(`COOKIES - ${userToken}`);
  const { user } = await getUser(ctx.cookies.get('userToken'));

  logtailClient.info(`USER NA APP - ${user}`);
  const request = {
    body: ctx.request.body,
    headers: ctx.req.headers,
    method: ctx.request.method,
    query: ctx.request.query
  };

  if (shouldRenderGraphiQL(request)) {
    ctx.body = renderGraphiQL({
      subscriptionsEndpoint: '/graphql'
    });
  } else {
    const { operationName, query, variables } = getGraphQLParameters(request);

    const result = await processRequest({
      operationName,
      query,
      variables,
      request,
      schema,
      contextFactory: () => {
        return getContext({
          req: request,
          user,
          koaContext: ctx as Context,
          setCookie: setCookie(ctx)
        });
      }
    });

    ctx.respond = false;
    sendResult(result, ctx.res);
  }
});

app.use(router.routes()).use(router.allowedMethods());

export default app;
