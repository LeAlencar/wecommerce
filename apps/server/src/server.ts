import { connectDB } from './database/database';
import dotenvSafe from 'dotenv-safe';
import { createServer } from 'http';
import app from './app';
import path from 'path';

(async () => {
  const cwd = process.cwd();

  const root = path.join.bind(cwd);
  dotenvSafe.config({
    path: root('.env'),
    sample: root('.env.example')
  });
  try {
    connectDB();
  } catch (error) {
    console.error('Unable to connect to database');
    process.exit(1);
  }

  const server = createServer(app.callback());

  server.listen(process.env.PORT, () => {
    console.log('Server running ğŸš€');
  });
})();
