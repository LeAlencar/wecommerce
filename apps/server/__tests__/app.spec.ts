import {
  clearDbAndRestartCounters,
  connectMongoose,
  defaultFrozenKeys,
  disconnectMongoose,
  sanitizeTestObject
} from '@wecommerce/test';
import { getContext } from '../src/getContext';
import { graphql } from 'graphql';
import { schema } from '../src/schema/schema';

it('should return 200 and show current version', async () => {
  const query = `
    query Q {
      version
    }
  `;

  const variables = {};

  const context = await getContext({});

  const rootValue = {};

  const result = await graphql({
    schema: schema,
    source: query,
    rootValue,
    contextValue: context,
    variableValues: variables
  });

  expect(result.data?.version).toBeTruthy();
  expect(
    sanitizeTestObject(result.data, [...defaultFrozenKeys])
  ).toMatchSnapshot();
});
