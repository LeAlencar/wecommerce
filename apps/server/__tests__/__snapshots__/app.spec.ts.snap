// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return 200 and return logged user 1`] = `
{
  "me": {
    "email": "user0@example.com",
    "id": "FROZEN-ID",
    "username": "user#0",
  },
}
`;
